src/query/sql/src/planner/semantic/type_check.rs

```diff
@@ -1098,6 +1123,85 @@ impl<'a> TypeChecker<'a> {
         }
     }
 
+    /// Resolve general window function call.
+    #[async_backtrace::framed]
+    async fn resolve_general_window_function(
+        &mut self,
+        func_name: &str,
+        args: &[&Expr],
+    ) -> Result<WindowFuncType> {
+        self.in_window_function = true;
+        let mut arguments = vec![];
+        let mut arg_types = vec![];
+        for arg in args.iter() {
+            let box (argument, arg_type) = self.resolve(arg).await?;
+            arguments.push(argument);
+            arg_types.push(arg_type);
+        }
+        self.in_window_function = false;
+        debug_assert!(!arguments.is_empty());
+        debug_assert!(!arg_types.is_empty());
+
+        let offset = if arguments.len() >= 2 {
+            let off = match &arguments[1] {
+                ScalarExpr::ConstantExpr(con) => match con.value {
+                    Scalar::Number(num) => match num {
+                        NumberScalar::UInt64(n) => Ok(n),
+                        NumberScalar::UInt32(n) => Ok(n as u64),
+                        NumberScalar::UInt16(n) => Ok(n as u64),
+                        NumberScalar::UInt8(n) => Ok(n as u64),
+                        _ => Err(ErrorCode::SemanticError(
+                            "Window function LAG/LEAD offset just support positive integer",
+                        )),
+                    },
+                    _ => Err(ErrorCode::SemanticError(
+                        "Window function LAG/LEAD offset just support positive integer",
+                    )),
+                },
+                _ => Err(ErrorCode::SemanticError(
+                    "Window function LAG/LEAD offset just support positive integer",
+                )),
+            };
+            Some(off?)
+        } else {
+            None
+        };
+
+        let default = if arguments.len() == 3 {
+            Some(arguments[2].clone())
+        } else {
+            None
+        };
+        let return_type = match default {
+            Some(_) => arg_types[0].clone(),
+            None => DataType::Nullable(Box::new(arg_types[0].clone())),
+        };
+
+        let cast_default = default
+            .map(|d| match d.clone() {
+                ScalarExpr::BoundColumnRef(_)
+                | ScalarExpr::CastExpr(_)
+                | ScalarExpr::ConstantExpr(_) => Ok(ScalarExpr::CastExpr(CastExpr {
+                    span: d.span(),
+                    is_try: true,
+                    argument: Box::new(d),
+                    target_type: Box::new(return_type.clone()),
+                })),
+                _ => Err(ErrorCode::SemanticError(
+                    "default value just support literal value and column, or ignore it",
+                )),
+            })
+            .transpose()?;
+
+        WindowFuncType::get_lag_lead_func(
+            func_name,
+            arguments[0].clone(),
+            offset,
+            cast_default,
+            return_type,
+        )
```

Comment:
```suggestion
        (WindowFuncType::Lag(LagLeadFunction {
              arg: Box::new(arguments[0].clone()),
              offset: offset.unwrap_or(1),
              default: cast_default.map(Box::new),
              return_type: Box::new(return_type),
          }
```

`get_lag_lead_func` this function is no need.